### PR TITLE
crush: cancel upmaps with up set size != pool size

### PR DESCRIPTION
**Pull request by hjwsm1989**
Fixes: https://tracker.ceph.com/issues/51842

Signed-off-by: huangjun <huangjun@xsky.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by hjwsm1989** [comment]
@tchaikov please help review.

**Comment by xiexingguo** [comment]
@hjwsm1989 Thanks for catching this bug.

Instead of doing it in verify_up, can you add a more explicit check for pg_upmap size at https://github.com/ceph/ceph/blob/master/src/osd/OSDMap.cc#L1960 ?

**Comment by hjwsm1989** [comment]
> @hjwsm1989 Thanks for catching this bug.
> 
> Instead of doing it in verify_up, can you add a more explicit check for pg_upmap size at https://github.com/ceph/ceph/blob/master/src/osd/OSDMap.cc#L1960 ?

@xiexingguo thanks for review, i think it's more efficient in verify_upmap, right?

**Comment by xiexingguo** [comment]
> thanks for review, i think it's more efficient in verify_upmap, right?

I don't think the efficiency matters much here.

My main concern is that there are various things (pg_temp/pg_upmap_items/pg_upmap/whatever) that can make up set go wrong, and it's always the preferred option if we can narrow the culprit down each time it happens.

**Comment by hjwsm1989** [comment]
> > thanks for review, i think it's more efficient in verify_upmap, right?
> 
> I don't think the efficiency matters much here.
> 
> My main concern is that there are various things (pg_temp/pg_upmap_items/pg_upmap/whatever) that can make up set go wrong, and it's always the preferred option if we can narrow the culprit down each time it happens.

em... that make sense, will do it ASAP.

**Comment by hjwsm1989** [comment]
@xiexingguo updated

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -1957,12 +1957,19 @@ bool OSDMap::check_pg_upmaps(
...
-      to_cancel->push_back(pg);
-      continue;
+    if (i != pg_upmap.end()) {
+      if (i->second == raw) {
+        ldout(cct, 10) << "removing redundant pg_upmap " << i->first << " "
+                       << i->second << dendl;
+        to_cancel->push_back(pg);
+        continue;
+      }
+      if ((int)i->second.size() > get_pg_pool_size(pg)) {
```
Comment (xiexingguo) [inline src/osd/OSDMap.cc:1967]: hmm, i think we should further narrow the condition down to **!=** to catch both oversize and undersize cases, e.g., we change pool size to a larger value after injecting some pg_upmaps into the osdmap.

**Review by xiexingguo — APPROVED** [review]
Thanks!

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-28_08:37:00-rados-wip-kefu-testing-2021-07-28-1257-distro-basic-smithi
- https://pulpito.ceph.com/kchai-2021-07-28_14:31:31-rados-wip-kefu-testing-2021-07-28-1908-distro-basic-smithi

failures tracked by 

- https://tracker.ceph.com/issues/49633

**Comment by tchaikov** [comment]
```
[ RUN      ] TestLibRBD.ConcurentOperations
using new format!
...
The following tests FAILED:
	 22 - run-rbd-unit-tests-61.sh (Timeout)
```

**Pull request merged**